common/helpers/src/avro.rs

```diff
@@ -0,0 +1,242 @@
+use crate::types::*;
+use apache_avro::{from_avro_datum, schema::Schema, to_avro_datum, types::Record, Codec, Writer};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use std::{collections::HashMap, io::Cursor, str};
+
+/// Represents error types returned by the `avro` module.
+#[derive(thiserror::Error, Debug)]
+pub enum AvroError {
+	#[error("I/O error")]
+	Io(#[from] std::io::Error),
+	#[error("Avro error")]
+	Avro(#[from] apache_avro::Error),
+	#[error("Invalid avro schema: {0}")]
+	InvalidSchema(String),
+	#[error("Invalid avro records")]
+	InvalidRecords(),
+}
+
+/// Function to convert a raw schema into serialized Avro schema.
+/// If schema is malformed or invalid, returns an error.
+/// # Arguments
+/// * `raw_schema` - raw schema to be converted
+/// # Returns
+/// * `Result<(Schema, Vec<u8>), AvroError>` - structured and serialized Avro schema
+/// # Examples
+/// ```
+/// use common_helpers::avro;
+/// use common_helpers::types::*;
+/// let raw_schema = r#"{"type": "record", "name": "User", "fields": [{"name": "name", "type": "string"}, {"name": "favorite_number", "type": "int"}]}"#;
+/// let schema_result = avro::fingerprint_raw_schema(raw_schema);
+/// assert!(schema_result.is_ok());
+/// let serialized_schema = schema_result.unwrap().1;
+pub fn fingerprint_raw_schema(raw_schema: &str) -> Result<(Schema, Vec<u8>), AvroError> {
+	// parse_str will fail if schema is not valid
+
+	let schema_result = Schema::parse_str(raw_schema)?;
+	let schema_canonical_form = schema_result.canonical_form();
+	// return the schema and the fingerprint
+	Ok((schema_result, schema_canonical_form.as_bytes().to_vec()))
+}
+
+/// Function to convert a list of raw schema into serialized Avro schema.
+/// If schema is malformed or invalid, it is set to Null.
+/// # Arguments
+/// * `raw_schema` - raw schema list to be converted
+/// # Returns
+/// * `Result<(Vec<Schema>, Vec<Vec<u8>>), AvroError>` - structured and serialized Avro schemas
+/// # Examples
+/// ```
+/// use common_helpers::avro;
+/// use common_helpers::types::*;
+/// let raw_schema = r#"{"type": "record", "name": "User", "fields": [{"name": "name", "type": "string"}, {"name": "favorite_number", "type": "int"}]}"#;
+/// let vec_raw_schema: [&str; 1] = [raw_schema];
+/// let schema_result = avro::fingerprint_raw_schema_list(&vec_raw_schema);
+/// assert!(schema_result.is_ok());
+/// let serialized_schemas = schema_result.unwrap().1;
+/// ```
+pub fn fingerprint_raw_schema_list(
+	raw_schema: &[&str],
+) -> Result<(Vec<Schema>, Vec<Vec<u8>>), AvroError> {
```

Comment:
when does this error?

Reply:
it errors when the function call on fingerprint_raw_schema errors